Sources/Queues/QueueWorker.swift

```diff
@@ -109,24 +111,73 @@ public struct QueueWorker {
                         self.queue.logger.error("Failed to send error notification: \(error)")
                         return self.queue.context.eventLoop.future()
+                    }.flatMap {
+                        logger.trace("Job done being run")
+                        return self.queue.clear(id)
                     }
                 }
             } else {
-                logger.error("Job failed, retrying... \(error)", metadata: [
-                    "job_id": .string(id.string),
-                    "job_name": .string(name),
-                    "queue": .string(self.queue.queueName.string)
-                ])
-                return self.run(
+                return self.retry(
                     id: id,
                     name: name,
                     job: job,
                     payload: payload,
                     logger: logger,
-                    remainingTries: remainingTries - 1,
-                    jobData: jobData
+                    remainingTries: remainingTries,
+                    attempts: attempts,
+                    jobData: jobData,
+                    error: error
                 )
             }
         }
     }
+
+    private func retry(
+            id: JobIdentifier,
+            name: String,
+            job: AnyJob,
+            payload: [UInt8],
+            logger: Logger,
+            remainingTries: Int,
+            attempts: Int?,
+            jobData: JobData,
+            error: Error
+    ) -> EventLoopFuture<Void> {
+        let attempts = attempts ?? 0
+        let delayInSeconds = job._nextRetryIn(attempt: attempts + 1)
+        if delayInSeconds == -1 {
+            logger.error("Job failed, retrying... \(error)", metadata: [
+                "job_id": .string(id.string),
+                "job_name": .string(name),
+                "queue": .string(self.queue.queueName.string)
+            ])
+            return self.run(
+                    id: id,
+                    name: name,
+                    job: job,
+                    payload: payload,
+                    logger: logger,
+                    remainingTries: remainingTries - 1 ,
+                    attempts: attempts + 1,
+                    jobData: jobData
+            )
+        } else {
+            logger.error("Job failed, retrying in \(delayInSeconds)s... \(error)", metadata: [
+                "job_id": .string(id.string),
+                "job_name": .string(name),
+                "queue": .string(self.queue.queueName.string)
+            ])
+            let storage = JobData(
+                    payload: jobData.payload,
+                    maxRetryCount: remainingTries - 1,
+                    jobName: jobData.jobName,
+                    delayUntil: Date(timeIntervalSinceNow: Double(delayInSeconds)),
+                    queuedAt: jobData.queuedAt,
+                    attempts: attempts + 1
+            )
+            return self.queue.set(id, to: storage).flatMap {
```

Comment:
Does this also remove the id from the processing queue?

Reply:
Run method use `pop()` so id will be taken out of the queue. I've assumed that `set` will overwrite jobs data if it is already stored, but probably it is safer to invoke `clear` before, so it won't break any of the queue drivers.